cql3/statements/select_statement.cc

```diff
@@ -1651,10 +1651,16 @@ void select_statement::check_needs_filtering(const restrictions::statement_restr
 {
     // non-key-range non-indexed queries cannot involve filtering underneath
     if (!_parameters->allow_filtering() && (restrictions.is_key_range() || restrictions.uses_secondary_indexing())) {
+        const auto& ck_restrictions = *restrictions.get_clustering_columns_restrictions();
+        const auto& pk_restrictions = *restrictions.get_partition_key_restrictions();
         // We will potentially filter data if either:
         //  - Have more than one IndexExpression
         //  - Have no index expression and the column filter is not the identity
-        if (restrictions.need_filtering()) {
+        if (restrictions.need_filtering()
+            // See #7608.
+            || ((pk_restrictions.empty() || has_token(pk_restrictions.expression))
+                && !ck_restrictions.empty() && !find_atom(ck_restrictions.expression, expr::is_multi_column)
+                && !restrictions.uses_secondary_indexing())) {
```

Comment:
and how does secondary indexes enter here?
I have to admit, I couldn't follow the logic of the if. Maybe you can update the comment above the if() to explain the new cases?

Reply:
> and how does secondary indexes enter here?

Because the unlimited-partitions check is on the base table.  If an index table is used, its partitions are always limited.

> Maybe you can update the comment above the if() to explain the new cases?

I'll try.

Reply:
Done.